Source/CarthageKit/Project.swift

```diff
@@ -409,10 +409,18 @@ public final class Project { // swiftlint:disable:this type_body_length
 	private func latestDependencies(resolver: ResolverProtocol) -> SignalProducer<[Dependency: PinnedVersion], CarthageError> {
 		return loadResolvedCartfile()
 			.map { $0.dependencies.mapValues { _ in VersionSpecifier.any } }
-			.flatMap(.merge) { resolver.resolve(dependencies: $0, lastResolved: nil, dependenciesToUpdate: nil) }
+			.flatMap(.merge) { dependencies -> SignalProducer<[Dependency: PinnedVersion], NoError> in
+				SignalProducer(dependencies)
+					.flatMap(.merge) { key, value -> SignalProducer<[Dependency: PinnedVersion], NoError> in
+						resolver
+							.resolve(dependencies: [key: value], lastResolved: nil, dependenciesToUpdate: nil)
+							.flatMapError { _ in .empty }
+					}
+					.reduce([Dependency: PinnedVersion]()) { $0.merging($1) { $1 } }
+		}
```

Comment:
Unfortunately, this isn't correct. This method should return the latest set of _compatible_ dependencies.

So if these versions are available:

- A
  - 1.0
  - 2.0
- B
  - 1.0: `A ~> 1.0`

Then it should return `A: 1.0, B: 1.0`, not `A: 2.0, B: 1.0`. The versions of the dependencies can't be considered in isolation.

I don't have an easy solution to this. Probably the easiest thing would be to:

1. Load the combined `Cartfile` in addition to `Cartfile.resolved`
2. If a dependency is specified with a `VersionSpecifier.gitReference`, then use it; otherwise use `.any` (what's currently being used)

That would prevent `outdated` from failing. It's probably not what we'd actually want, which would be to still show versions if possible.

Or… maybe it would be better to:

1. `resolve`
2. catch `CarthageError.requiredVersionNotFound`
    1. pin the erring dependency
    2. Go to (1)

That would be slower in the worst case, but seems like it would return the correct results in all cases. 🤔

Reply:
Well, I think this method is to "Attempts to determine the latest version (whether satisfiable or not)" as written here https://github.com/Carthage/Carthage/pull/2328/files#diff-3bd7f6198aa606c5ec6931e77cb3c42bR404. Isn't it correct?

Reply:
Ah, I misunderstood this method ignores version compatibility😱 I'll fix it.

Reply:
Addressed https://github.com/Carthage/Carthage/pull/2328/commits/9d96115fe235c414e72876ba3fb9d7cf5357c052.